File: SarahIA/SarahIA/Services/TranslationEngine.swift
import Foundation

/// Moteur de Traduction Multilingue Résilient (Français, Hébreu, Anglais) :
/// - Dictionnaire local ultra-rapide (0ms, 100% hors-ligne)
/// - Requête réseau MyMemory avec timeout court (4s) et fallback local automatique
/// - Compatible iOS 12.0+ (Callbacks) et iOS 13.0+ (Async/Await)
public final class TranslationEngine {
    
    public static let shared = TranslationEngine()
    
    public enum TargetLanguage: String {
        case french = "fr"
        case hebrew = "he"
        case english = "en"
        
        public var displayNameFr: String {
            switch self {
            case .french: return "Français"
            case .hebrew: return "Hébreu"
            case .english: return "Anglais"
            }
        }
    }
    
    public struct TranslationRequest {
        public let textToTranslate: String
        public let sourceLanguage: String
        public let targetLanguage: TargetLanguage
    }
    
    private lazy var session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 4.0
        config.timeoutIntervalForResource = 6.0
        config.requestCachePolicy = .returnCacheDataElseLoad
        return URLSession(configuration: config)
    }()
    
    private let fastLocalDict: [String: String] = [
        // FR -> HE
        "bonjour": "שלום",
        "salut": "היי",
        "merci": "תודה",
        "merci beaucoup": "תודה רבה",
        "au revoir": "להתראות",
        "s'il vous plaît": "בבקשה",
        "s'il te plaît": "בבקשה",
        "oui": "כן",
        "non": "לא",
        "bonne nuit": "לילה טוב",
        "bonsoir": "ערב טוב",
        "comment ça va": "מה נשמע",
        "comment vas tu": "מה שלומך",
        "je t'aime": "אני אוהב אותך",
        "bienvenue": "ברוכים הבאים",
        "bon appétit": "בתיאבון",
        "félicitations": "מזל טוב",
        "pardon": "סליחה",
        
        // HE -> FR
        "שלום": "Bonjour",
        "תודה": "Merci",
        "תודה רבה": "Merci beaucoup",
        "להתראות": "Au revoir",
        "בבקשה": "S'il vous plaît",
        "כן": "Oui",
        "לא": "Non",
        "לילה טוב": "Bonne nuit",
        "ערב טוב": "Bonsoir",
        "מה נשמע": "Comment ça va ?",
        "מה שלומך": "Comment vas-tu ?",
        "סליחה": "Pardon / Excusez-moi",
        
        // FR -> EN
        "bonjour|en": "Hello",
        "salut|en": "Hi",
        "merci|en": "Thank you",
        "merci beaucoup|en": "Thank you very much",
        "au revoir|en": "Goodbye",
        "comment ça va|en": "How are you?",
        "bonne nuit|en": "Good night",
        "je t'aime|en": "I love you"
    ]
    
    private init() {}
    
    /// Analyse la phrase pour détecter une demande explicite de traduction
    public func parseTranslationIntent(input: String) -> TranslationRequest? {
        let lower = input.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        
        // 1. Vers l'hébreu
        if lower.contains("en hébreu") || lower.contains("en hebreu") {
            let clean = lower
                .replacingOccurrences(of: "traduis", with: "")
                .replacingOccurrences(of: "traduit", with: "")
                .replacingOccurrences(of: "comment on dit", with: "")
                .replacingOccurrences(of: "comment se dit", with: "")
                .replacingOccurrences(of: "comment dit on", with: "")
                .replacingOccurrences(of: "en hébreu", with: "")
                .replacingOccurrences(of: "en hebreu", with: "")
                .replacingOccurrences(of: ":", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !clean.isEmpty {
                return TranslationRequest(textToTranslate: clean, sourceLanguage: "fr", targetLanguage: .hebrew)
            }
        }
        
        // 2. Vers l'anglais
        if lower.contains("en anglais") {
            let clean = lower
                .replacingOccurrences(of: "traduis", with: "")
                .replacingOccurrences(of: "traduit", with: "")
                .replacingOccurrences(of: "comment on dit", with: "")
                .replacingOccurrences(of: "comment se dit", with: "")
                .replacingOccurrences(of: "en anglais", with: "")
                .replacingOccurrences(of: ":", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !clean.isEmpty {
                return TranslationRequest(textToTranslate: clean, sourceLanguage: "fr", targetLanguage: .english)
            }
        }
        
        // 3. Vers le français
        if lower.contains("en français") || lower.contains("en francais") || lower.contains("לצרפתית") || lower.contains("בצרפתית") {
            let clean = lower
                .replacingOccurrences(of: "traduis", with: "")
                .replacingOccurrences(of: "traduit", with: "")
                .replacingOccurrences(of: "תרגם", with: "")
                .replacingOccurrences(of: "איך אומרים", with: "")
                .replacingOccurrences(of: "en français", with: "")
                .replacingOccurrences(of: "en francais", with: "")
                .replacingOccurrences(of: "לצרפתית", with: "")
                .replacingOccurrences(of: "בצרפתית", with: "")
                .replacingOccurrences(of: ":", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !clean.isEmpty {
                let src = detectLanguage(text: clean) == "he" ? "he" : "en"
                return TranslationRequest(textToTranslate: clean, sourceLanguage: src, targetLanguage: .french)
            }
        }
        
        return nil
    }
    
    public func detectLanguage(text: String) -> String {
        var hebrewCount = 0
        var latinCount = 0
        for scalar in text.unicodeScalars {
            let val = scalar.value
            if val >= 0x0590 && val <= 0x05FF {
                hebrewCount += 1
            } else if (val >= 65 && val <= 90) || (val >= 97 && val <= 122) || (val >= 0x00C0 && val <= 0x017F) {
                latinCount += 1
            }
        }
        return (hebrewCount > latinCount && hebrewCount > 0) ? "he" : "fr"
    }
    
    // MARK: - Traduction avec Callback (iOS 12.0+)
    
    public func translate(text: String, sourceLang: String, targetLang: TargetLanguage, completion: @escaping (String) -> Void) {
        let clean = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let lower = clean.lowercased()
        
        // 1. Dictionnaire local instantané (0ms, 100% hors-ligne)
        let dictKey = targetLang == .english ? "\(lower)|en" : lower
        if let localMatch = fastLocalDict[dictKey] {
            completion(localMatch)
            return
        }
        
        // 2. Requête API distante sécurisée avec timeout 4s
        guard let encoded = clean.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "https://api.mymemory.translated.net/get?q=\(encoded)&langpair=\(sourceLang)|\(targetLang.rawValue)") else {
            completion(clean)
            return
        }
        
        let request = URLRequest(url: url)
        session.dataTask(with: request) { data, response, error in
            if let data = data,
               let http = response as? HTTPURLResponse, http.statusCode == 200,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let responseData = json["responseData"] as? [String: Any],
               let translatedText = responseData["translatedText"] as? String,
               !translatedText.isEmpty,
               !translatedText.lowercased().contains("no query specified") {
                completion(translatedText)
            } else {
                // Fallback sur le texte d'origine
                completion(clean)
            }
        }.resume()
    }
    
    // MARK: - Traduction Asynchrone (iOS 13.0+)
    
    @available(iOS 13.0, *)
    public func translate(text: String, sourceLang: String, targetLang: TargetLanguage) async -> String {
        return await withCheckedContinuation { continuation in
            self.translate(text: text, sourceLang: sourceLang, targetLang: targetLang) { translated in
                continuation.resume(returning: translated)
            }
        }
    }
}

File: android/app/src/main/java/com/sarahia/app/TranslationEngine.kt
package com.sarahia.app

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.Executors
import java.util.regex.Pattern

/**
 * Moteur de Traduction Multilingue Temps Réel (Français, Hébreu, Anglais) :
 * - Traduction bidirectionnelle instantanée FR ⇄ HE, FR ⇄ EN, EN ⇄ FR
 * - Détection automatique de la langue
 * - Dictionnaire local ultra-rapide (< 1ms) + moteur distant haute précision
 */
class TranslationEngine(private val context: Context) {

    private val TAG = "SarahTranslation"
    private val executor = Executors.newFixedThreadPool(2)

    // Dictionnaire local haute fréquence pour des réponses à la milliseconde
    private val fastLocalDict = mapOf(
        // FR -> HE
        "bonjour" to "שלום",
        "salut" to "היי",
        "merci" to "תודה",
        "merci beaucoup" to "תודה רבה",
        "au revoir" to "להתראות",
        "s'il vous plaît" to "בבקשה",
        "s'il te plaît" to "בבקשה",
        "oui" to "כן",
        "non" to "לא",
        "bonne nuit" to "לילה טוב",
        "bonsoir" to "ערב טוב",
        "bonne journée" to "יום טוב",
        "comment ça va" to "מה נשמע",
        "comment vas tu" to "מה שלומך",
        "je t'aime" to "אני אוהב אותך",
        "bienvenue" to "ברוכים הבאים",
        "bon appétit" to "בתיאבון",
        "félicitations" to "מזל טוב",
        "pardon" to "סליחה",
        "excusez moi" to "סליחה",
        "à bientôt" to "נתראה בקרוב",

        // HE -> FR
        "שלום" to "Bonjour",
        "תודה" to "Merci",
        "תודה רבה" to "Merci beaucoup",
        "להתראות" to "Au revoir",
        "בבקשה" to "S'il vous plaît",
        "כן" to "Oui",
        "לא" to "Non",
        "לילה טוב" to "Bonne nuit",
        "ערב טוב" to "Bonsoir",
        "יום טוב" to "Bonne journée",
        "מה נשמע" to "Comment ça va ?",
        "מה שלומך" to "Comment vas-tu ?",
        "סליחה" to "Pardon / Excusez-moi",
        "מזל טוב" to "Félicitations",

        // FR -> EN
        "bonjour|en" to "Hello",
        "merci|en" to "Thank you",
        "au revoir|en" to "Goodbye",
        "comment ça va|en" to "How are you?",
        "bonne nuit|en" to "Good night",
        "s'il vous plaît|en" to "Please"
    )

    enum class TargetLanguage(val code: String, val displayNameFr: String, val displayNameHe: String) {
        FRENCH("fr", "Français", "צרפתית"),
        HEBREW("he", "Hébreu", "עברית"),
        ENGLISH("en", "Anglais", "אנגלית")
    }

    data class TranslationRequest(
        val textToTranslate: String,
        val sourceLanguage: String,
        val targetLanguage: TargetLanguage
    )

    /**
     * Détecte si la phrase de l'utilisateur est une demande explicite de traduction.
     */
    fun parseTranslationIntent(input: String): TranslationRequest? {
        val lower = input.trim().lowercase()

        // 1. Vers l'hébreu
        val toHebrewPatterns = listOf(
            Pattern.compile("^(?:traduis|traduit|traduire|comment dit on|comment on dit|comment se dit)\\s+(?:en hébreu|en hebreu)\\s*[:,-]?\\s*(.+)$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(.+)\\s+en hébreu\\s*\\??$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(?:traduis|traduit|traduire)\\s+(.+)\\s+(?:en hébreu|en hebreu)$", Pattern.CASE_INSENSITIVE)
        )
        for (p in toHebrewPatterns) {
            val m = p.matcher(lower)
            if (m.find()) {
                val text = m.group(1)?.trim() ?: ""
                if (text.isNotEmpty()) return TranslationRequest(text, "fr", TargetLanguage.HEBREW)
            }
        }

        // 2. Vers l'anglais
        val toEnglishPatterns = listOf(
            Pattern.compile("^(?:traduis|traduit|traduire|comment dit on|comment on dit|comment se dit)\\s+(?:en anglais)\\s*[:,-]?\\s*(.+)$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(.+)\\s+en anglais\\s*\\??$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(?:traduis|traduit|traduire)\\s+(.+)\\s+(?:en anglais)$", Pattern.CASE_INSENSITIVE)
        )
        for (p in toEnglishPatterns) {
            val m = p.matcher(lower)
            if (m.find()) {
                val text = m.group(1)?.trim() ?: ""
                if (text.isNotEmpty()) return TranslationRequest(text, "fr", TargetLanguage.ENGLISH)
            }
        }

        // 3. Vers le français
        val toFrenchPatterns = listOf(
            Pattern.compile("^(?:traduis|traduit|traduire|comment dit on|comment on dit)\\s+(?:en français|en francais)\\s*[:,-]?\\s*(.+)$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(.+)\\s+en français\\s*\\??$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(?:תרגם|איך אומרים|איך מתרגמים)\\s+(?:לצרפתית|בצרפתית)\\s*[:,-]?\\s*(.+)$", Pattern.CASE_INSENSITIVE)
        )
        for (p in toFrenchPatterns) {
            val m = p.matcher(lower)
            if (m.find()) {
                val text = m.group(1)?.trim() ?: ""
                val src = if (detectLanguage(text) == "he") "he" else "en"
                if (text.isNotEmpty()) return TranslationRequest(text, src, TargetLanguage.FRENCH)
            }
        }

        return null
    }

    fun detectLanguage(text: String): String {
        var hebrewCount = 0
        var latinCount = 0
        for (char in text) {
            val code = char.code
            if (code in 0x0590..0x05FF) {
                hebrewCount++
            } else if ((code in 65..90) || (code in 97..122) || code in 0x00C0..0x017F) {
                latinCount++
            }
        }
        return if (hebrewCount > latinCount && hebrewCount > 0) "he" else "fr"
    }

    /**
     * Exécute la traduction de manière asynchrone ultra-rapide.
     */
    fun translateAsync(
        text: String,
        sourceLang: String,
        targetLang: TargetLanguage,
        callback: (Result<String>) -> Unit
    ) {
        executor.execute {
            val cleanText = text.trim()
            val lower = cleanText.lowercase()

            // 1. Vérification dans le dictionnaire local instantané
            val dictKey = if (targetLang == TargetLanguage.ENGLISH) "$lower|en" else lower
            val localMatch = fastLocalDict[dictKey]
            if (localMatch != null) {
                callback(Result.success(localMatch))
                return@execute
            }

            // 2. Traduction via service haute fidélité (MyMemory / LibreTranslate)
            try {
                val pair = "${sourceLang}|${targetLang.code}"
                val encodedText = URLEncoder.encode(cleanText, "UTF-8")
                val urlString = "https://api.mymemory.translated.net/get?q=$encodedText&langpair=$pair"
                
                val url = URL(urlString)
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "GET"
                conn.connectTimeout = 4000
                conn.readTimeout = 4000

                if (conn.responseCode == 200) {
                    val reader = BufferedReader(InputStreamReader(conn.inputStream, "UTF-8"))
                    val resp = reader.readText()
                    reader.close()

                    val json = JSONObject(resp)
                    val responseData = json.optJSONObject("responseData")
                    val translatedText = responseData?.optString("translatedText", "") ?: ""

                    if (translatedText.isNotEmpty() && !translatedText.equals("NO QUERY SPECIFIED", ignoreCase = true)) {
                        callback(Result.success(translatedText))
                        return@execute
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "Distant translation failed: ${e.message}")
            }

            // Fallback de secours
            callback(Result.success("« $cleanText » (${targetLang.displayNameFr})"))
        }
    }
}
